src/core/border_router/routing_manager.cpp

```diff
@@ -1326,17 +1322,79 @@ bool RoutingManager::UpdateRouterAdvMessage(const RouterAdv::RouterAdvMessage *a
     if (aRouterAdvMessage == nullptr || aRouterAdvMessage->GetRouterLifetime() == 0)
     {
         mRouterAdvMessage.SetToDefault();
+        mLearntRouterAdvMessageFromHost = false;
     }
     else
     {
-        mRouterAdvMessage = *aRouterAdvMessage;
-        mDiscoveredPrefixStaleTimer.FireAtIfEarlier(mTimeRouterAdvMessageLastUpdate +
-                                                    Time::SecToMsec(kRtrAdvStaleTime));
+        mRouterAdvMessage               = *aRouterAdvMessage;
+        mLearntRouterAdvMessageFromHost = true;
     }
 
+    ResetDiscoveredPrefixStaleTimer();
+
     return (mRouterAdvMessage != oldRouterAdvMessage);
 }
 
+void RoutingManager::ResetDiscoveredPrefixStaleTimer(void)
+{
+    TimeMilli now                           = TimerMilli::GetNow();
+    TimeMilli nextStaleTime                 = now.GetDistantFuture();
+    TimeMilli maxOnlinkPrefixStaleTime      = now;
+    bool      requireCheckStaleOnlinkPrefix = false;
+
+    OT_ASSERT(mIsRunning);
+
+    // The stale timer is responding for sending RS to re-check the stale On-Link/OMR prefixes or RA message.
+    // The rules for calculating the next stale time:
+    // 1. If BR learns RA header from Host daemons, it should send RS when the RA header is stale.
+    // 2. If BR discovered any on-link prefix, it should send RS when all on-link prefixes are stale.
+    // 3. If BR discovered any OMR prefix, it should send RS when the first OMR prefix is stale.
+
+    // Check for stale Router Advertisement Message if learnt from Host.
+    if (mLearntRouterAdvMessageFromHost)
+    {
+        TimeMilli routerAdvMessageStaleTime = mTimeRouterAdvMessageLastUpdate + Time::SecToMsec(kRtrAdvStaleTime);
+
+        nextStaleTime = OT_MIN(nextStaleTime, routerAdvMessageStaleTime);
+    }
+
+    for (ExternalPrefix &externalPrefix : mDiscoveredPrefixes)
+    {
+        TimeMilli prefixStaleTime = externalPrefix.GetStaleTime();
+
+        if (externalPrefix.mIsOnLinkPrefix && !externalPrefix.IsDeprecated())
+        {
+            // Check for least recent stale On-Link Prefixes if BR is not advertising local On-Link Prefix.
+            maxOnlinkPrefixStaleTime      = OT_MAX(maxOnlinkPrefixStaleTime, prefixStaleTime);
+            requireCheckStaleOnlinkPrefix = true;
+        }
+        else if (!externalPrefix.mIsOnLinkPrefix)
+        {
+            // Check for most recent stale OMR Prefixes
+            nextStaleTime = OT_MIN(nextStaleTime, prefixStaleTime);
+        }
```

Comment:
nit: avoid an additional if check
```suggestion
        if (externalPrefix.mIsOnLinkPrefix)
        {
            if (!externalPrefix.IsDeprecated()))
            {
                // Check for least recent stale On-Link Prefixes if BR is not advertising local On-Link Prefix.
                maxOnlinkPrefixStaleTime      = OT_MAX(maxOnlinkPrefixStaleTime, prefixStaleTime);
                requireCheckStaleOnlinkPrefix = true;
            }
        }
        else
        {
            // Check for most recent stale OMR Prefixes
            nextStaleTime = OT_MIN(nextStaleTime, prefixStaleTime);
        }
```